### PR TITLE
[WIP] Add the HierarchicalSoftmax activation function to torch.nn #634

### PR DESCRIPTION
**Pull request by cooljackal**
Working on adding HierarchicalSoftmax to torch.nn using the existing Softmax function as a guide. The original issue is #634. Initially there may be only one (default) option for tree creation. The current task list is:

- [x] Add cloned HierarchicalSoftmax class to torch.nn.modules.activation (and __all__ export)
- [x] Add HierarchicalSoftmax import to torch.nn.modules.__init__ (and __all__ export)
- [x] Add hierarchical_softmax function to torch.nn.functional
- [x] Add static Huffman tree generation function
- [ ] Add unit tests for Huffman tree
- [ ] Update docstrings to define base/optional functionality
- [ ] Update C++ kernel (optional)

Notes:
- hierarchical_softmax in torch.nn.functional uses the same unary, dim, and dtype checks as softmax

### CODE REVIEW

**Comment by albanD** [comment]
Will let @jbschlosser take care of this one, feel free to add me back if needed though.

**Comment by jbschlosser** [comment]
Thanks @cooljackal for working on this! Happy to review this when the full impl & unit tests are added. If you need any help with where to put things, etc. feel free to ping me :)

**Comment by cooljackal** [comment]
> Thanks @cooljackal for working on this! Happy to review this when the full impl & unit tests are added. If you need any help with where to put things, etc. feel free to ping me :)

No problem! I may have to take you up on the offer to answer some questions later on. Right now I am still working on the approach, tracing with gdb, and testing different tree generation implementations.

**Comment by cooljackal** [comment]
I am hitting some local build issues suddenly on my PC, so having to spend time working through what happened. I have potentially eliminated the conda environment, the git repo itself (downloaded it again), various env vars (reduced to only disabling CUDA/ROCM, ccache, ninja, and different branches/master as issues. It seems to be breaking on caffe2 with namespace errors for some reason, even if caffe2 is disabled.

**Comment by cooljackal** [comment]
Fixed the build issues. I am going to do a feature branch off release/1.12 instead of the current master commit.

**Comment by jbschlosser** [comment]
> Fixed the build issues. I am going to do a feature branch off release/1.12 instead of the current master commit.

Probably best to keep it off of master, as we won't be accepting PRs to release/1.12

**Comment by mruberry** [comment]
@albanD would you review this?

**Comment by cooljackal** [comment]
@mruberry I still have a few more items that I need to add, hoping to have those items added over the next few weeks. I just needed to rebase with a large set of changes from master.
